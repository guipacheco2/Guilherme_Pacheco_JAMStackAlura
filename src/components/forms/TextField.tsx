import React from 'react'
import styled, { css } from 'styled-components'
import { Typography, TypographyPropsGeneric } from '../foundation'

const StyledTextField = styled.div`
  margin: 8px 0;
`

const StyledInput = styled(Typography)(({ theme }) => {
  return css`
    width: 100%;
    border: 1px solid
      ${theme.schema === 'light'
        ? theme.colors.primary.main
        : theme.colors.primary.contrastText};
    color: ${theme.colors.surface.contrastText};
    background-color: ${theme.schema === 'light'
      ? theme.colors.surface.main
      : theme.colors.primary.dark};
    padding: 12px 16px;
    margin: 4px 0;
    outline: 0;
  `
}) as React.ComponentType<TypographyPropsGeneric<'input'>>

const StyledTextArea = styled(Typography)(({ theme }) => {
  return css`
    width: 100%;
    border: 1px solid
      ${theme.schema === 'light'
        ? theme.colors.primary.main
        : theme.colors.primary.contrastText};
    color: ${theme.colors.surface.contrastText};
    background-color: ${theme.schema === 'light'
      ? theme.colors.surface.main
      : theme.colors.primary.dark};
    padding: 12px 16px;
    margin: 4px 0;
    outline: 0;
    resize: none;
  `
}) as React.ComponentType<TypographyPropsGeneric<'textarea'>>

interface TextFieldProps {
  id: string
  label: string
  placeholder: string
  name: string
  type?: string
  multiline?: boolean
  onChange?: (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
  ) => void
  value: string
}

export function TextField({
  id,
  label,
  type = 'text',
  placeholder,
  name,
  multiline,
  onChange,
  value,
}: TextFieldProps): JSX.Element {
  return (
    <StyledTextField>
      <Typography
        as="label"
        htmlFor={id}
        variant="bodyText1"
        onColor="background"
      >
        {label}
      </Typography>

      {multiline ? (
        <StyledTextArea
          as="textarea"
          placeholder={placeholder}
          name={name}
          onChange={onChange}
          value={value}
          variant="bodyText1"
          id={id}
          onColor="primary"
          rows={3}
        />
      ) : (
        <StyledInput
          as="input"
          type={type}
          id={id}
          placeholder={placeholder}
          name={name}
          onChange={onChange}
          value={value}
          variant="bodyText1"
          onColor="primary"
        />
      )}
    </StyledTextField>
  )
}
