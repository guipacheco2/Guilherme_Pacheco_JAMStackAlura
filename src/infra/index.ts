export * from './hooks'
export * from './test'
