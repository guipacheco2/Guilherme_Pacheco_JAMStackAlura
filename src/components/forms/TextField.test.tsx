import user from '@testing-library/user-event'
import React from 'react'
import { render, screen } from '../../infra'
import { TextField } from './TextField'

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        value="Ju"
        onChange={() => false}
        name="nome"
        id="nome"
        label="Nome"
      />,
    )

    const textField = screen.getByPlaceholderText(/nome/i)
    expect(textField).toMatchSnapshot()
  })

  describe('and user is typing', () => {
    test('the value must be updated', () => {
      const onChangeMock = jest.fn()

      render(
        <TextField
          placeholder="Nome"
          value=""
          onChange={onChangeMock}
          name="nome"
          id="nome"
          label="Nome"
        />,
      )

      const inputNome = screen.getByPlaceholderText(/nome/i)
      const typedName = 'someusername'
      user.type(inputNome, typedName)
      expect(onChangeMock).toHaveBeenCalledTimes(typedName.length)
    })
  })
})
