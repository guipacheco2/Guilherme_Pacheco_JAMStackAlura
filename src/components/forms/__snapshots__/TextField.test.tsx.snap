// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  width: 100%;
  border: 1px solid #607D8B;
  color: #34515E;
  background-color: #FAFAFA;
  padding: 12px 16px;
  margin: 4px 0;
  outline: 0;
}

<input
  class="c0"
  id="nome"
  name="nome"
  placeholder="Nome"
  type="text"
  value="Ju"
/>
`;
